fixing save checkpoint for  best metric

makifozkanoglu opened the pull request:
## Motivation
I trying save best weight while validating the data by using evaluation hook script. However, if save_best argument is true, it could not handle, gives following error

'''
exec(compile(contents + "\n", file, 'exec'), glob, loc)
  File "tools/train.py", line 202, in <module>
    main()
  File "tools/train.py", line 197, in main
    meta=meta,
  File "/root/**/ssod/apis/train.py", line 206, in train_detector
    runner.run(data_loaders, cfg.workflow)
  File "/root/**/thirdparty/mmcv/mmcv/runner/epoch_based_runner.py", line 127, in run
    epoch_runner(data_loaders[i], **kwargs)
  File "/root/**/thirdparty/mmcv/mmcv/runner/epoch_based_runner.py", line 54, in train
    self.call_hook('after_train_epoch')
  File "/root/**/thirdparty/mmcv/mmcv/runner/base_runner.py", line 309, in call_hook
    getattr(hook, fn_name)(self)
  File "/root/**/thirdparty/mmcv/mmcv/runner/hooks/evaluation.py", line 267, in after_train_epoch
    self._do_evaluate(runner)
  File "/root/**/ssod/utils/hooks/evaluation.py", line 57, in _do_evaluate
    self._save_ckpt(runner, key_score)
  File "/root/**/thirdparty/mmcv/mmcv/runner/hooks/evaluation.py", line 347, in _save_ckpt
    self.out_dir, best_ckpt_name, create_symlink=False)
TypeError: save_checkpoint() takes 2 positional arguments but 3 were given
'''
    

## Modification


runner.save_checkpoint method needs a an argument named as filename_tmpl, and filename_tmpl should include '{}' in order for following line to work without error

 line 213, iter_based_runner.py: filename = filename_tmpl.format(self.iter + 1)
line 166, epoch_based_runner.py: filename = filename_tmpl.format(self.epoch + 1)


Therefore line 347 is edited as shown

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/open-mmlab/mmcv?pullRequest=1694) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/open-mmlab/mmcv?pullRequest=1694) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/open-mmlab/mmcv?pullRequest=1694) it.</sub>

zhouzaida commented:
Hi, when the `filename_tmpl` do not contain `{}`, the `filename_tmpl.format(x)` should equal to `filename_tmpl`.

```python
>>> filename_tmpl = 'hello '
>>> filename_tmpl.format('world')
'hello '
```

makifozkanoglu commented:
Thanks I got it is not necessary to include {}, but why does it give following TypeError in two different computer. In my request, is not it a solution by indicating second argurement's name like filename_tmpl=f'best_{self.key_indicator}_' + cur_type + '_{}.pth', Currently, I have no problem in torch distributed launcher

TypeError: save_checkpoint() takes 2 positional arguments but 3 were given

makifozkanoglu commented:
What is your final suggestion, what should i do for that error,

zhouzaida commented:
Could you provide some steps to reproduce the error?

makifozkanoglu commented:
I attached my config file below. Besides, the file in the link ( [click](https://github.com/microsoft/SoftTeacher/blob/bef9a256e5c920723280146fc66b82629b3ee9d4/tools/train.py))  is used as training scripts

Here is the starting command
python -m torch.distributed.launch --nproc_per_node 1 --master_port 29500 tools/train.py configs/myconfig.py --launcher pytorch
```
mmdet_base = "../thirdparty/mmdetection/configs/_base_"

_base_ = [
    f"{mmdet_base}/datasets/coco_detection.py",
    f"{mmdet_base}/schedules/schedule_1x.py", 
    f"{mmdet_base}/default_runtime.py"
]

model = dict(
    type='CenterNet',
    backbone=dict(
        type='HourglassNet',
        # depth=18,
        # norm_eval=False,
        # norm_cfg=dict(type='BN'),
        # init_cfg=dict(type='Pretrained', checkpoint='torchvision://resnet18')
        ),
    # neck=None,
    neck=dict(
        type='CTResNetNeck',
        in_channel=256,
        num_deconv_kernels=(4, 4, 4),
        num_deconv_filters=(256, 128, 64),
        use_dcn=False
        ),
    bbox_head=dict(
        type='CenterNetHead',
        num_classes=80,
        in_channel=64,
        feat_channel=64,
        loss_center_heatmap=dict(type='GaussianFocalLoss', loss_weight=1.0),
        loss_wh=dict(type='L1Loss', loss_weight=0.1),
        loss_offset=dict(type='L1Loss', loss_weight=1.0)),
    train_cfg=None,
    test_cfg=dict(topk=100, local_maximum_kernel=3, max_per_img=100))

# We fixed the incorrect img_norm_cfg problem in the source code.
img_norm_cfg = dict(
    mean=[123.675, 116.28, 103.53], std=[58.395, 57.12, 57.375], to_rgb=True)

train_pipeline = [
    dict(type='LoadImageFromFile', to_float32=True, color_type='color'),
    dict(type='LoadAnnotations', with_bbox=True),
    dict(
        type='PhotoMetricDistortion',
        brightness_delta=32,
        contrast_range=(0.5, 1.5),
        saturation_range=(0.5, 1.5),
        hue_delta=18),
    dict(
        type='RandomCenterCropPad',
        crop_size=(512, 512),
        ratios=(0.6, 0.7, 0.8, 0.9, 1.0, 1.1, 1.2, 1.3),
        mean=[0, 0, 0],
        std=[1, 1, 1],
        to_rgb=True,
        test_pad_mode=None),
    dict(type='Resize', img_scale=(512, 512), keep_ratio=True),
    dict(type='RandomFlip', flip_ratio=0.5),
    dict(type='Normalize', **img_norm_cfg),
    dict(type='DefaultFormatBundle'),
    dict(type='Collect', keys=['img', 'gt_bboxes', 'gt_labels'])
]
test_pipeline = [
    dict(type='LoadImageFromFile', to_float32=True),
    dict(
        type='MultiScaleFlipAug',
        scale_factor=1.0,
        flip=False,
        transforms=[
            dict(type='Resize', keep_ratio=True),
            dict(
                type='RandomCenterCropPad',
                ratios=None,
                border=None,
                mean=[0, 0, 0],
                std=[1, 1, 1],
                to_rgb=True,
                test_mode=True,
                test_pad_mode=['logical_or', 31],
                test_pad_add_pix=1),
            dict(type='RandomFlip'),
            dict(type='Normalize', **img_norm_cfg),
            dict(type='DefaultFormatBundle'),
            dict(
                type='Collect',
                meta_keys=('filename', 'ori_shape', 'img_shape', 'pad_shape',
                           'scale_factor', 'flip', 'flip_direction',
                           'img_norm_cfg', 'border'),
                keys=['img'])
        ])
]


dataset_type = 'CocoDataset'
data_root =  'data/coco/'

# Use RepeatDataset to speed up training
data = dict(
    samples_per_gpu=4,
    workers_per_gpu=4,
    train=dict(
         # _delete_=True,
        #  type='RepeatDataset',
        #times=5,
        #dataset=dict(
            type=dataset_type,
            ann_file=data_root + 'instances_train2017.json',
            img_prefix=data_root + 'train',
           #  classes=classes,
            pipeline=train_pipeline #)
        ),
    val=dict(type=dataset_type,
             ann_file=data_root + 'instances_val2017.json',
             img_prefix=data_root + 'val',
             # classes=classes,
             pipeline=test_pipeline),
    # test=dict(pipeline=test_pipeline),
    test=dict(type=dataset_type,
              ann_file=data_root + 'instances_val2017.json',
              img_prefix=data_root + 'val',
              # classes=classes,
              pipeline=test_pipeline),
    sampler=dict(
        train=dict(
            type="Sampler",
            # ind_unused_parameters=True,
            # samples_per_gpu=8
            # sample_ratio=[1],
            # by_prob=True,
            # at_least_one=True,
            # epoch_length=7330,
        )
    ),
)

# optimizer

optimizer = dict(type='SGD', lr=2.5e-4, momentum=0.9, weight_decay=0.0001) #0.02
optimizer_config = dict(
    _delete_=True, grad_clip=dict(max_norm=35, norm_type=2))

# learning policy
# Based on the default settings of modern detectors, we added warmup settings.
lr_config = dict(
    policy='step',
    warmup='linear',
    warmup_iters=1000,
    warmup_ratio=1.0 / 1000,
    step=[18*5, 24*5])  # the real step is [18*5, 24*5]
runner = dict(max_epochs=28*10)  # the real epoch is 28*5=140

# checkpoint_config = dict(interval=10, by_epoch=True)
checkpoint_config = dict(type='CustomCheckpointHook', 
                         interval=1, 
                         max_keep_ckpts=1, 
                         by_epoch=True)

workflow = [('train', 1), ('val', 1)]
log_config = dict(
    interval=0,
    hooks=[
        dict(type="TextLoggerHook", interval=0,by_epoch=True,ignore_last=False),
    ],
)

evaluation = dict(interval=1,
                  metric='bbox',
                  save_best='auto'
                  )
work_dir = "work_dirs/CenterNetVHR_HG4

"
```

makifozkanoglu commented:
I noticed that the error may occur because of the method named as patch_checkpoint(runner: BaseRunner) in  [patch.py](https://github.com/microsoft/SoftTeacher/blob/bef9a256e5c920723280146fc66b82629b3ee9d4/ssod/utils/patch.py).

Because it reinitialise the save_checkpoint method of runner similar to python decorators. 
FYI: the methods in patch.py are included in [apis/train.py](https://github.com/microsoft/SoftTeacher/blob/bef9a256e5c920723280146fc66b82629b3ee9d4/ssod/apis/train.py) file

zhouzaida commented:
Thank you for the detailed information. As you have described, the error was exactly caused by the `patch_checkpoint`.

zhouzaida commented on mmcv/runner/hooks/evaluation.py:
```diff
@@ -344,7 +344,10 @@ def _save_ckpt(self, runner, key_score):
             runner.save_checkpoint(
-                self.out_dir, best_ckpt_name, create_symlink=False)
+                self.out_dir, 
+                filename_tmpl=f'best_{self.key_indicator}_' + cur_type + '_{}.pth', 
```
Should we directly pass the `best_ckpt_name` to the `filename_tmpl`?
```suggestion
                filename_tmpl=best_ckpt_name, 
```

makifozkanoglu replied:
Yes we could, but, might including iter number in the filename be a good option?In my opinion I would prefer to see iteration number in the filename

zhouzaida replied:
The original `best_ckpt_name`  has contained the current epoch or iteration.

makifozkanoglu commented:
will you approve the PR?

zhouzaida commented:
> will you approve the PR?

Yet, I will merge the PR after CI passed. Thanks for your contribution very much.

zhouzaida commented:
Hi, you can fix this line https://github.com/open-mmlab/mmcv/blob/b83bdb0cbee98c5e73ff193042a832381565d313/tests/test_ops/test_roi_align_rotated.py#L83 with `from mmcv.ops import RoIAlignRotated, roi_align_rotated`

zhouzaida commented on mmcv/runner/hooks/evaluation.py:
```diff
@@ -344,7 +344,10 @@ def _save_ckpt(self, runner, key_score):
-                self.out_dir, best_ckpt_name, create_symlink=False)
+                self.out_dir,
+                filename_tmpl=best_ckpt_name,
+                create_symlink=False
+            )
```
```suggestion
                create_symlink=False)
```
